CommonTools/PileupAlgos/plugins/PuppiPhoton.cc

```diff
@@ -186,6 +186,7 @@ void PuppiPhoton::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
 }
 // ------------------------------------------------------------------------------------------
 bool PuppiPhoton::matchPFCandidate(const reco::Candidate *iPF,const reco::Candidate *iPho) { 
+  if(iPF->pdgId() != iPho->pdgId()) return false;
   double lDR = deltaR(iPF->eta(),iPF->phi(),iPho->eta(),iPho->phi());
   for(unsigned int i0 = 0; i0 < pdgIds_.size(); i0++) {
     if(std::abs(iPF->pdgId()) == pdgIds_[i0] && lDR < dRMatch_[i0])  return true;
```

Comment:
the matching is to make sure it is the vary same PF candidate, so the same PDG ID is required. So electrons can be matched to electrons and photons to photons, but not mixed.
The loop over pdg id allows to configure different dR criteria for different particle types.